core-v1/src/main/scala/data/Justified.scala

```diff
@@ -2,34 +2,126 @@ package com.rallyhealth.vapors.v1
   def value: V
-  def withValue[U](value: U): Justified[U]
 
-  def evidenceOrNone: Evidence = Evidence.none
+  def reason: String
+
+  def visit[G[+_]](v: Justified.Visitor[G]): G[V]
 
-  final def map[U](fn: V => U): Justified[U] = withValue(fn(value))
+  def configs: Seq[(String, Option[String])] // TODO: How should this handle duplicates?
+  def evidence: Evidence
+
+  def zipWith[Y, Z](
+    that: Justified[Y],
+    reason: String,
+  )(
+    fn: (V, Y) => Z,
+  ): Justified[Z] =
+    Justified.ByInference(
+      reason,
+      fn(this.value, that.value),
+      NonEmptyList.of(this, that),
+    )
 }
 
 object Justified {
 
-  final case class Constant[+V](value: V) extends Justified[V] {
-    override def withValue[U](value: U): Justified[U] = copy(value = value)
+  def apply[V](value: V): Justified[V] = ByConst(value)
+
+  trait Visitor[G[+_]] {
+
+    def visitConstant[V](justified: ByConst[V]): G[V]
+
+    def visitConfig[V](justified: ByConfig[V]): G[V]
+
+    def visitFact[V](justified: ByFact[V]): G[V]
+
+    def visitInference[V](justified: ByInference[V]): G[V]
   }
 
-  final case class ByConfig[+V](
+  def byConst[V](value: V): Justified[V] = ByConst(value)
+
+  final case class ByConst[V](value: V) extends Justified[V] {
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConstant(this)
+    override def evidence: NoEvidence.type = NoEvidence
+    override val reason: String = s"[const: $value]"
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByConst"
+  }
+
+  def byConfig[V](
+    value: V,
+    configKey: String,
+    configDescription: Option[String] = None,
+  ): Justified[V] = ByConfig(value, configKey, configDescription)
+
+  final case class ByConfig[V](
     value: V,
     configKey: String,
     configDescription: Option[String] = None,
   ) extends Justified[V] {
-    override def withValue[U](value: U): Justified[U] = copy(value = value)
+    override def evidence: NoEvidence.type = NoEvidence
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConfig(this)
+    override val reason: String = s"[config: $configKey=$value]"
+    override def configs: Seq[(String, Option[String])] = Seq(configKey -> configDescription)
+    override def productPrefix: String = "Justified.ByConfig"
   }
 
-  final case class ByEvidence[+V](
+  def byFact[V](fact: TypedFact[V]): Justified[V] = ByFact(fact)
+
+  // TODO: Allow a lens for map operations?
+  final case class ByFact[V](fact: TypedFact[V]) extends Justified[V] {
+    override val evidence: SomeEvidence = SomeEvidence(NonEmptySet.of(fact))
+    override def value: V = fact.value
+    override val reason: String = s"[fact: ${fact.typeInfo.name}=${fact.value}]"
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitFact(this)
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByFact"
+  }
```

Comment:
I achieve the same effect of a map operation using the `ByInference` class... so I'm not sure if this TODO is needed. Best argument I can make in favor of adding the lens here is that it might reduce the depth of the justification tree because all `valuesOfType` operations return a sequence of justified values by facts, and most operations on these individual facts will start with a `.get` selector (unless the fact is already the correct type of value, in which case the lens will be the identity lens). I'm not sure it is worth it.

I should either leave a more detailed comment for future experimentation or just remove the comment and add it back if needed.